### ISSUE TITLE
How to unblock a user using privacy list

### ISSUE THREAD

**Issue by Shahjahan786** [QuickBlox/quickblox-android-sdk#542]
I'm using following code to block user which is successful and also code to unblock the user[code is given below] but it gives exception if privacy list has no item [last item removed by unblocking].
also it gives exception on calling   `privacyListsManager.deletePrivacyList(SharedPreferenceUtils.getCurrentUserID());`

public static boolean blockUser(int userId) {

        boolean userBlocked = false;

        QBPrivacyList privacyList = null;

        getPrivacyListsManager();

        if (privacyListsManager == null) {
            AppUtils.logDebug("privacyListsManager", "privacyListsManager is null");
            return false;
        }

        try {
            privacyList = privacyListsManager.getPrivacyList(SharedPreferenceUtils.getCurrentUserID());
        } catch (Exception e) {
            e.printStackTrace();
        }

        if (privacyList == null) {
            privacyList = new QBPrivacyList();
            privacyList.setName(SharedPreferenceUtils.getCurrentUserID());
        }

        QBPrivacyListItem item1 = new QBPrivacyListItem();
        item1.setType(QBPrivacyListItem.Type.USER_ID);
        item1.setAllow(false);
        item1.setMutualBlock(true);
        item1.setValueForType(String.valueOf(userId));

        ArrayList<QBPrivacyListItem> items = new ArrayList<>();

        if (null != privacyList.getItems()) {
            items.addAll(privacyList.getItems());
        }

        boolean bFound = false;
        for (QBPrivacyListItem listItem : items) {
            if (listItem.getValueForType().equals(item1.getValueForType())) {
                listItem.setAllow(false);
                item1.setMutualBlock(true);
                bFound = true;
                break;
            }
        }

        if (!bFound) {
            items.add(item1);
        }
        AppUtils.logDebug("PRIVACY_LIST[ITEMS]", items.toString());
        privacyList.setItems(items);

        try {
            privacyListsManager.declineActiveList();
        } catch (SmackException.NotConnectedException e) {
            e.printStackTrace();
        } catch (XMPPException.XMPPErrorException e) {
            e.printStackTrace();
        } catch (SmackException.NoResponseException e) {
            e.printStackTrace();
        }

        try {
            privacyListsManager.setPrivacyList(privacyList);
            MyApplication.privacyListItems = items;
            userBlocked = true;
        } catch (Exception e) {
            userBlocked = false;
            e.printStackTrace();
        }

        try {
        
            privacyListsManager.setPrivacyListAsActive(SharedPreferenceUtils.getCurrentUserID());

        } catch (Exception e) {
            e.printStackTrace();
        }
        return userBlocked;
    }

    public static boolean unBlockUser(int userId) {

        boolean userUnBlocked = false;

        QBPrivacyList privacyList = null;

        getPrivacyListsManager();

        if (privacyListsManager == null) {
            AppUtils.logDebug("privacyListsManager", "privacyListsManager is null");
            return false;
        }

        try {
            privacyList = privacyListsManager.getPrivacyList(SharedPreferenceUtils.getCurrentUserID());
        } catch (Exception e) {
            e.printStackTrace();
            return userUnBlocked;
        }

        List<QBPrivacyListItem> items = MyApplication.privacyListItems;

//        if (null != privacyList.getItems()) {
//            items.addAll(privacyList.getItems());
//        }

        for (int i = 0; i < items.size(); i++) {
            QBPrivacyListItem listItem = items.get(i);
            if (listItem.getValueForType().contains(String.valueOf(userId))) {
                items.remove(i);
                break;
            }
        }

        AppUtils.logDebug("PRIVACY_LIST[ITEMS]", items.toString());
        privacyList.setItems(items);

        try {
            privacyListsManager.declineActiveList();
            if(items.isEmpty()){
                privacyListsManager.deletePrivacyList(SharedPreferenceUtils.getCurrentUserID());
                NayaPayApplication.privacyListItems = items;
                userUnBlocked = true;
            }

        } catch (SmackException.NotConnectedException e) {
            e.printStackTrace();
        } catch (XMPPException.XMPPErrorException e) {
            e.printStackTrace();
        } catch (SmackException.NoResponseException e) {
            e.printStackTrace();
        }

        try {
            if(items != null && !items.isEmpty()) {
                privacyListsManager.setPrivacyList(privacyList);
                NayaPayApplication.privacyListItems = items;
                userUnBlocked = true;
            }
        } catch (Exception e) {
            userUnBlocked = false;
            e.printStackTrace();
        }

        try {
            privacyListsManager.setPrivacyListAsDefault(SharedPreferenceUtils.getCurrentUserID());
            privacyListsManager.setPrivacyListAsActive(SharedPreferenceUtils.getCurrentUserID());

        } catch (Exception e) {
            e.printStackTrace();
        }
        return userUnBlocked;
    }

    public static boolean isUserBlocked(int userId) {
        boolean isUserBlocked = false;
        List<QBPrivacyListItem> items = MyApplication.privacyListItems;
        for (int i = 0; i < items.size(); i++) {
            QBPrivacyListItem listItem = items.get(i);

            if (listItem.getValueForType().contains(String.valueOf(userId))) {
                isUserBlocked = true;
                break;
            }
        }

        return isUserBlocked;
    }

    public static List<QBPrivacyListItem> getPrivacyListItems() {
        List<QBPrivacyListItem> privacyListItems = new ArrayList<>();

        QBPrivacyList privacyList = null;

        getPrivacyListsManager();

        if (privacyListsManager == null) {
            AppUtils.logDebug("privacyListsManager", "privacyListsManager is null");
            return privacyListItems;
        }

        try {
            privacyList = privacyListsManager.getPrivacyList(SharedPreferenceUtils.getCurrentUserID());

            if (privacyList != null) {
                privacyListItems = privacyList.getItems();

                if (privacyListItems != null && !privacyListItems.isEmpty()) {
                    NayaPayApplication.privacyListItems = privacyListItems;
                }

            }

        } catch (Exception e) {
            e.printStackTrace();
            return privacyListItems;
        }

        return privacyListItems;
    }

**LOG**

https://gist.github.com/Shahjahan786/82607e104dc4c4ab105acb5f315542b3

**Comment by Shahjahan786** [comment]
Is privacylist applied per user[logged in] or per application?

**Comment by tatanka987** [comment]
please see [documentation for privacy lists](https://quickblox.com/developers/Android_XMPP_Chat_Sample#Delete_existing_privacy_list), before deleting privacy list you should decline it then delete

**Comment by Shahjahan786** [comment]
I have read documentation and As you can see the code snippet I have already called  `privacyListsManager.declineActiveList();` before calling `privacyListsManager.deletePrivacyList(SharedPreferenceUtils.getCurrentUserID()`

**Comment by tatanka987** [comment]
As you can see error packet has message "Can not modify list while being in use by other session". It means you can't modify privacy list when same user logged on several devices. 
Privacy list is set for user.

**Comment by tatanka987** [comment]
You can't change Active privacy list when same user logged on several devices. But on newest version of SDK we use only one privacy list (Default) without Active. Which version of SDK do you use?

**Comment by Shahjahan786** [comment]
I use SDK 3.5.0

**Comment by Shahjahan786** [comment]
Any update on this?

**Comment by tatanka987** [comment]
1) privacy list works for current user only, you can't create 'public' privacy list.
2)-
3)no, you shouldn't, you set active list one time and it will work when you will login to the chat.
4) 'default' privacy list works all time (except way when you logged in to the chat and sed 'active' list). We recommend  use 'default' privacy list without 'active' because it will simplify your app and it will work all time (not only when user logged in to the chat). 'active' privacy list works only whet user logged in to the chat.
5) Use [this snippet ](https://quickblox.com/developers/Android_XMPP_Chat_Sample#Create_a_privacy_list_or_edit_existing_list). To block user use `item1.setAllow(false);`, to unblock use `item1.setAllow(true);` or remove this item from list.

**Comment by Shahjahan786** [comment]
Thanks @tatanka987 

I'm now clear about this but want to know that should I set name of my privacylist as below
```
QBPrivacyList list = new QBPrivacyList();
list.setName("public");
```

or 
QBPrivacyList list = new QBPrivacyList();
list.setName("name_of_each_user");

**Comment by tatanka987** [comment]
you can use any name for your privacy list (e.g. 'default' or 'active', or 'userId', or 'userLogin')

**Comment by Shahjahan786** [comment]
What I have understood is:
suppose privacyListName = 'my_privacy_list' (default list - I set it for all users in application)
when user A blocks user B
user C block User A

now User B is blocked only for user A and allowed for User C
same way User A is blocked for User C and allowed for B
-- User A,B,C are allowed for all other users in application

**Comment by tatanka987** [comment]
yes, you are correct

**Comment by Shahjahan786** [comment]
Thanks a lot @tatanka987

**Comment by QB-nikolay-zolotarov** [comment]
Hello QuickBlox customer,

This is Nikolay from QuickBlox support.

The issue was closed as it was outdated.

Please check the relevant section of our documentation here:
[https://docs.quickblox.com/docs/android-chat-privacy-list](https://docs.quickblox.com/docs/android-chat-privacy-list)

Also, please update the SDK to the latest version:
[https://github.com/QuickBlox/quickblox-android-sdk-releases/releases/tag/3.9.11](https://github.com/QuickBlox/quickblox-android-sdk-releases/releases/tag/3.9.11)

Additionally, please check our new samples:
[https://docs.quickblox.com/docs/code-samples#chat-samples](https://docs.quickblox.com/docs/code-samples#chat-samples)

If it is still relevant after reviewing the updated information, feel free to open a new issue.

Have a nice day.
